### PR TITLE
WT-5677 Prepare support with durable history: add test/format stress tests

### PR DESCRIPTION
**Pull request by keithbostic**
@ddanderson, here are changes that I think get you most of the way to where you're heading, lightly tested.

The one thing I specifically have not tested is that making the thread operations structures long-lived actually allows reads to work across worker thread restart, I'm not sure if reads are all failing, or whatever. I figured that would be something you'd have to check anyway, so I just left it alone.

If you like this approach but want changes, let me know if I can help, or please feel free to simply discard this work without further discussion if you want to go in a different direction.

### CODE REVIEW

**Comment by agorrod** [comment]
@ddanderson let us know if you'd like anything more here.

**Comment by ddanderson** [comment]
@keithbostic, @agorrod The state of this test/format with this PR is that if `transaction.rollback_to_stable=1` on the command line, it gives an `EBUSY` error when calling `WT_SESSION->rebalance`.  Not sure yet if this is an error in WT or the test.  In this change, `transaction.rollback_to_stable` will not be selected to be on randomly, so the failure is hidden.  Is it worth reviewing/merging and opening a different ticket for the rebalance issue?  I will investigate in the meantime.

**Comment by kommiharibabu** [comment]
@keithbostic As part of recovery, rollback to stable the stable timestamp is set in `__wt_txn_recover()` and it never reset. This is one place that stable timestamp is changed.

```
        /*
         * Set the stable timestamp from recovery timestamp and process the trees for rollback to
         * stable.
         */
        conn->txn_global.stable_timestamp = conn->txn_global.recovery_timestamp;
        conn->txn_global.has_stable_timestamp = true;
```
The stable timestamp doesn't reset like oldest timestamp because verify operation expects the stable timestamp is expected to be set. 

I am not sure whether this may be the reason behind the problem.

**Comment by ddanderson** [comment]
Thanks, @keithbostic and @kommiharibabu for tracking that.  I think it's appropriate to spin that issue off in another ticket, I'll do that.  In the meantime, can you both please review this PR?

Keith, there are only a few mods from what you sent me.  It seems to be working well from what I can tell.  You're the author of the PR, so I can't explicitly mark you as a reviewer.

Hari, if you can focus on whether the DH operations make sense and provide a reasonable test: basically, test/format has a thread that advances the `oldest_timestamp`, now (if `rollback_to_stable` is configured) it moves the `stable_timestamp` in lock-step.  The main loop starts a bunch of threads, then all threads stop, then (if configured) `rollback_to_stable` is called, and we go back to the top of the loop.  After a few of these, it runs a bunch of utility functions including verify, salvage.  The worker threads of course, do usual CRUD operations + prepare (if configured) and use time stamps if configured.

**Comment by ddanderson** [comment]
I've pasted your comments into the ticket https://jira.mongodb.org/browse/WT-6043 where the discussion about the rebalance issue can be continued.

File: test/format/config.h [hunk 1]
```diff
@@ -329,6 +329,9 @@ static CONFIG c[] = {
     "isolation level (random | read-uncommitted | read-committed | snapshot)", C_IGNORE | C_STRING,
     0, 0, 0, NULL, &g.c_isolation},
 
+  {"transaction.rollback_to_stable", "if rollback_to_stable should be called", C_BOOL, 0, 0, 0,
+    &g.c_txn_rollback_to_stable, NULL},
+
```
Comment (keithbostic) [inline test/format/config.h:334]: There's no base percentage to configure this, was that deliberate? That is, this won't ever be set unless it's set on the command line?

File: test/format/t.c [hunk 2]
```diff
@@ -283,13 +285,24 @@ main(int argc, char *argv[])
...
 
         /* Operations. */
-        for (reps = 1; reps <= FORMAT_OPERATION_REPS; ++reps)
-            operations(ops_seconds, reps == FORMAT_OPERATION_REPS);
+        for (reps = 1; reps <= FORMAT_OPERATION_REPS; ++reps) {
+            last_run = (reps == FORMAT_OPERATION_REPS);
+            operations(ops_seconds, last_run);
+            if (!last_run && g.c_txn_rollback_to_stable)
+                g.wts_conn->rollback_to_stable(g.wts_conn, NULL);
+        }
```
Comment (keithbostic) [inline test/format/t.c:305]: Why don't we do rollback-to-stable on the last run?

File: test/format/util.c [hunk 3]
```diff
@@ -253,10 +253,22 @@ timestamp_once(WT_SESSION *session)
...
+        /*
+         * When we're doing rollback to stable operations, we'll advance the stable timestamp too,
+         * in lockstep with the oldest timestamp.
+         */
+        if (g.c_txn_rollback_to_stable) {
+            len = strlen(buf);
+            WT_ASSERT((WT_SESSION_IMPL *)session, len < sizeof(buf));
+            testutil_check(
+              __wt_snprintf(buf + len, sizeof(buf) - len, ",%s%s", stable_timestamp_str, tsbuf));
+        }
```
Comment (keithbostic) [inline test/format/util.c:269]: Why do we move the stable timestamp in lockstep with the oldest timestamp, wouldn't they typically be different values in real applications?

**Comment by kommiharibabu** [comment]
@ddanderson I reviewed the changes but I don't have much knowledge of `test/format`. I didn't find that after performing rollback to stable, are we verifying whether the data present is stable or not? Is it done already, if not it is better to do?

**Comment by ddanderson** [comment]
I'm closing this PR, I'm going to put its work for rollback_to_stable into a different PR for ticket WT-6108. Meanwhile, I just started PR #5598, which will contain some minimal work just for prepare ( WT-5677).

**Pull request closed without merging**